src/nxt_http_request.c

```diff
@@ -283,7 +283,15 @@ nxt_http_request_create(nxt_task_t *task)
     task->thread->engine->requests_cnt++;
 
     r->tstr_cache.var.pool = mp;
-
+#if (NXT_HAVE_OTEL)
+    if (nxt_otel_is_init()) {
```

Comment:
In this case `nxt_otel_init` is run during configuration (see changes to nxt_router.c).
Every request will have its own unique state, or if uninitialized the `request->otel` pointer will remain NULL. 

The NULL case for this pointer is checked in nxt_otel_test_and_call_state (nxt_otel.c).

Reply:
Suppose there are three configurations because of the reconfiguration, how many otel instances will there be?

Reply:
The otel instance is one per request. It is mostly made from fields parsed in the reequest header, or autogenerated independently of the tracer configuration. The router thread may reconfigure many times but the request object otel state will not be touched.

Reply:
> Suppose there are three configurations because of the reconfiguration, how many otel instances will there be?

I mean independent otel instance but not r->otel which is a pointer to a otel instance.

Reply:
The otel instances are not allocated by the opentelemetry configuration. They are only allocated once when the request begins processing.

Configuration will call directly the rust code `nxt_otel_init` and `nxt_otel_uninit` to reinitialize the global tracer. In this design spans (linked to by the otel state instance) that have been processed but not sent should then use the updated global tracer.

Is your concern that multiple reconfigurations will result in requests having additional otel instances allocated?

Reply:
As I understand for requests with the same host, they should be traced or not, right?
But because of the reconfiguration, it's possible that most of the requests will be traced, but parts of them are not.
The Unit's reconfiguration is to keep the same behavior for requests while the configuration is changed at times.

I thought is that, for configuration A, there is otel instance if option is enabled, for configuration B, there is also an individual instance if option is enabled. Let me know if I misunderstand the intention.

Reply:
During normal Unit work, when a request / language work / response is still in flight and Unit gets reconfigured, the Unit process that's dealing with the request according to the old configuration isn't getting shut down immediately, right? Like, it waits until the response is sent, connection is finished, and closed, and then the process is cleaned up?

Reply:
Almost correct.

We call the process that deals with incoming request `router process`, and the router process has serval threads: one main thread and other threads that deal with the incoming request.
The process is always alive and isn't affected by the reconfiguration, and the same for the threads unless you change the thread numbers.

> Like, it waits until the response is sent, connection is finished, and closed, and then the process is cleaned up?

then the old configuration is cleaned up, I mean free it from allocated memory.

Make it short, while the reconfiguration happens, the alive requests pointer to the old configuration, and the new requests will pointer to the new configuration.

Reply:
Okay, so the otel configuration is alive in the router process, which means even if we reconfigure it, and the new configuration does not have otel enabled, the old one in flight will still finish correctly because the the otel bits are alive as long as the request is.

Reply:
Yep, and @ac000 mentioned about what the exact enable mean, it’s related to compile option, configuration option like settings which is global scope and specific router action option, it belongs to the requirements topic.